### PR TITLE
Resolve barrier mismatch error for multilocale libraries run with more than one locale

### PR DESCRIPTION
**Pull request by lydia-duncan**
We observed that multilocale libraries would behave fine when only
one locale was used, but that using multiple locales would encounter
a gasnet barrier mismatch error per locale when cleaning up the library.

By having a separate main function which called chpl_mli_smain, instead of
altering chpl_gen_main, we were skipping some crucial bits of information and
accidentally running chpl_mli_smain once per locale.  This change gets us to
use the normal strategy for calling chpl_gen_main (which is only called once
per program) and then has that call chpl_mli_smain.

In order to call chpl_mli_smain, we needed to get the connection information
out of argv (where the client library put it).  Add a function to the internal
modules, similar to chpl_convert_args, which gets the connection information
so that it can be passed to chpl_mli_smain.

No longer link to main_mli, instead just to the regular main (since that is what
calls chpl_gen_main).  Also, change chpl_mli_smain to accept a `const char*`
instead of just a `char*` since that is the type of `c_string` under the covers
and we need to use `c_string` for chpl_gen_main to call it.

Alter the runtime's handling of the connection argument to actually
store it in chpl_main_argument, since that's how we communicate argv/argc to
chpl_gen_main.  Remove now unused main_mli.c and rules for building it.

While here, I also removed some extraneous space from main.c.

Add Python test to lock in that using more than one locale works.  I also verified
that a C test works for this (but the commit for that is in the C testing branch),
and that it works for both `--static` and `--dynamic` compilation.

Testing:
- passed local Mac testing of the python multilocale interop tests
- passed linux64 single locale with futures
- no new failures on linux64 gasnet with futures (leftover failures from the List module, but those are present on master)

### CODE REVIEW

**Comment by lydia-duncan** [comment]
@dlongnecke-cray - I'd like to merge this today if possible, will you be able to review?

**Comment by dlongnecke-cray** [comment]
I have no idea how I missed such an important thread in my email, I'll do that right now!

File: test/interop/python/multilocale/checkMultipleLocales.good [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Hello, world! (from task 0 of 2 on locale 0 of 4)
```
Comment (dlongnecke-cray) [inline test/interop/python/multilocale/checkMultipleLocales.good]: The order of the output for this test is always guaranteed to be this, even though it's running in a `coforall`?

Comment (lydia-duncan) [reply test/interop/python/multilocale/checkMultipleLocales.good]: Yup, I added a prediff that sorts the output

**Pull request merged**